### PR TITLE
fix(gdn): honor Q/K normalization in GDN prefill

### PR DESCRIPTION
**Pull request by akaashrp**
## 📌 Description

`chunk_gated_delta_rule` ignored `use_qk_l2norm_in_kernel=True` on the native prefill paths. Callers passing unnormalized Q/K could get incorrect results or NaNs. This PR normalizes Q/K before dispatch when the flag is enabled. The CuTe DSL kernel handles both tensors in one launch, accumulates in FP32 with epsilon `1e-6`, and casts back to the input dtype. Calls with the flag disabled keep their existing behavior.

## 🔍 Related Issues

None.

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [ ] I have installed the hooks with `pre-commit install`.
- [ ] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

Pre-commit passed on all three changed files, including mypy and Ruff. The local commit hook runs in Docker. I haven't run `pre-commit run --all-files`.

```bash
pre-commit run --files flashinfer/gdn_prefill.py \
  flashinfer/gdn_kernels/qk_l2norm.py tests/gdn/test_prefill_qk_normalization.py
```

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [x] All tests are passing (`unittest`, etc.).

```bash
python -m pytest tests/gdn/test_prefill_qk_normalization.py -q
```

- **H200 (SM90):** 10 passed.
- **B100 (SM100)** 10 passed.

The tests cover FP16/BF16, GQA/GVA, variable-length sequences, nonzero initial states, CP/non-CP paths, and CUDA graphs. Normalization is checked against an FP64 reference. I also ran Qwen3.5-4B on H200 at 512, 4,096, and 32,768 tokens with the normalization flag enabled. The original code produced nonfinite logits at each length. With this fix, all three runs produced finite logits.

## 🔬 Experimental Track

- [ ] This PR is **experimental**: it adds or changes code under `flashinfer/experimental/` and/or an `@flashinfer_experimental_api`. Tracking issue: #
  - [ ] The tracking issue names an owner, the reason for the experimental path, and a graduation plan with a target release.
  - [ ] Core changes are limited to a thin entry point (signature, shared validation, feature-gate check, backend selection, handoff).
  - [ ] Tests live in `tests/experimental/` and were validated on the intended hardware; a runnable example is included.
  - [ ] Nothing is registered in `flashinfer/aot.py`, and no experimental backend is reachable from `backend="auto"` without `FLASHINFER_ALLOW_EXPERIMENTAL_AUTO_BACKENDS=1`. (Calling an `@flashinfer_experimental_api` or naming a backend explicitly is itself the opt-in and needs no environment variable.)
  - [ ] **Test scope declared below.** The experimental CI lane runs exactly these targets, so keep them as narrow as the change allows.

```experimental-tests
# One target per line: a directory or a file. (A pytest ::selector is not
# supported -- the sharding runner cannot consume one.) Must be under
# tests/experimental/ and must exist. Delete these comment lines and add yours, e.g.
#
#   tests/experimental/test_my_backend.py
#   tests/experimental/my_backend/
#
# Declaring the whole tree (tests/experimental/) is allowed but means every
# experimental PR pays for every other feature's tests, in every matrix cell.
```

## Summary by CodeRabbit

* **New Features**
  * Added Q/K L2 normalization support for gated delta rule prefill operations on supported CUDA architectures.
  * Normalization uses FP32 accumulation and an epsilon for numerical stability while preserving the input data type.
  * Supports Q and K tensors with different row counts, including empty inputs.
  * Unsupported hardware and configurations continue to receive appropriate validation errors.

* **Tests**
  * Added coverage across data types, head shapes, context-parallel configurations, unsupported hardware, and CUDA graph replay determinism.

### CODE REVIEW

**Comment by akaashrp** [comment]
@flashinfer-bot run

**Comment by akaashrp** [comment]
@flashinfer-bot rerun

**Comment by akaashrp** [comment]
@flashinfer-bot run

**Comment by akaashrp** [comment]
@flashinfer-bot run

**Comment by akaashrp** [comment]
@flashinfer-bot run

**Comment by akaashrp** [comment]
@flashinfer-bot run

**Comment by akaashrp** [comment]
@flashinfer-bot run

**Review by jhjpark — APPROVED** [review]
